pkg/controller/elasticsearch/elasticsearch_controller_test.go

```diff
@@ -0,0 +1,261 @@
+					},
+				},
+			},
+			false,
+			newBuilder("testESwithtoolongofanamereallylongname", "test").
+				WithGeneration(2).
+				WithAnnotations(map[string]string{hints.OrchestrationsHintsAnnotation: `{"no_transient_settings":false}`}).
+				WithStatus(esv1.ElasticsearchStatus{ObservedGeneration: 2, Phase: esv1.ElasticsearchResourceInvalid}).BuildAndCopy(),
+		},
+		{
+			"ES with too long name, and needing annotations update, fails initial reconcile, and does not have status.* updated because of a 409/resource conflict",
+			k8sClientFields{
+				false,
+				nil,
+				[]runtime.Object{
+					newBuilder("testESwithtoolongofanamereallylongname", "test").
+						WithGeneration(2).
+						WithStatus(esv1.ElasticsearchStatus{ObservedGeneration: 1}).Build()},
+			},
+			args{
+				request: reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      "testESwithtoolongofanamereallylongname",
+						Namespace: "test",
+					},
+				},
+			},
+			false,
+			newBuilder("testESwithtoolongofanamereallylongname", "test").
+				WithGeneration(2).
+				WithAnnotations(map[string]string{hints.OrchestrationsHintsAnnotation: `{"no_transient_settings":false}`}).
+				WithStatus(esv1.ElasticsearchStatus{ObservedGeneration: 1}).BuildAndCopy(),
+		},
+		{
+			"Invalid ES version errors, and updates observedGeneration",
+			k8sClientFields{
+				false,
+				nil,
+				[]runtime.Object{
+					newBuilder("testES", "test").
+						WithGeneration(2).
+						WithVersion("invalid").
+						WithAnnotations(map[string]string{hints.OrchestrationsHintsAnnotation: `{"no_transient_settings":false}`}).
+						WithStatus(esv1.ElasticsearchStatus{ObservedGeneration: 1}).Build()},
+			},
+			args{
+				request: reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      "testES",
+						Namespace: "test",
+					},
+				},
+			},
+			false,
+			newBuilder("testES", "test").
+				WithGeneration(2).
+				WithVersion("invalid").
+				WithAnnotations(map[string]string{hints.OrchestrationsHintsAnnotation: `{"no_transient_settings":false}`}).
+				WithStatus(esv1.ElasticsearchStatus{ObservedGeneration: 2, Phase: esv1.ElasticsearchResourceInvalid}).BuildAndCopy(),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := newTestReconciler(tt.k8sClientFields.failing, tt.k8sClientFields.err, tt.k8sClientFields.objects...)
+			if _, err := r.Reconcile(context.Background(), tt.args.request); (err != nil) != tt.wantErr {
+				t.Errorf("ReconcileElasticsearch.Reconcile() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			// get the actual ES from the k8s client, ensuring the k8s client is not a failing client
+			// comparing to the expected.
+			var actualES esv1.Elasticsearch
+			ensureClient(r, tt.k8sClientFields.failing, tt.k8sClientFields.objects)
+			if err := r.Client.Get(context.Background(), tt.args.request.NamespacedName, &actualES); err != nil {
+				t.Error(err)
+				return
+			}
```

Comment:
Ok, I've allows the failing k8s client to be disabled/enabled on demand, instead of generating a new client, which better tests the intended behavior.